library/ssl_msg.c

```diff
@@ -3195,10 +3290,17 @@ int mbedtls_ssl_prepare_handshake_record( mbedtls_ssl_context *ssl )
     }
     else
 #endif /* MBEDTLS_SSL_PROTO_DTLS */
-    /* With TLS we don't handle fragmentation (for now) */
+#if defined(MBEDTLS_SSL_HS_DEFRAG_CLT)
+    if( mbedtls_ssl_hs_defrag_enabled( ssl ) )
+    {
+        if( ( ret = ssl_defrag_try_read( ssl, ssl->in_hslen, TRY_READ_FLAG_COMMIT ) ) != 0 )
```

Comment:
Committing prior to processing the content returned from the reader is -- strictly speaking -- an API violation. The promise of the reader API is that the buffers returned from it are valid _until_ the data is committed. 

I'd suggest to remove the FLAG parameter from `ssl_defrag_try_read()` and instead commit as part of `consume_message()`.

Reply:
Addressed in the latest update.